docs/design-docs/01-oci-support.md

```diff
@@ -0,0 +1,418 @@
+
+```yaml
+upstream:
+  type: oci
+  oci:
+    image: us-docker.pkg.dev/the-project-id/the-repo-name/the-package:v14
+```
+
+To update to an upstream digest, `kpt pkg update @sha256:{SHA256_HEX}` and `kpt pkg update DIR@sha256:{SHA256_HEX}` will assign `@sha256:{SHA256_HEX}` as the new upstream image digest.
+
+```yaml
+upstream:
+  type: oci
+  oci:
+    image: us-docker.pkg.dev/the-project-id/the-repo-name/the-package@sha256:8815143a333cb9d2cb341f10b984b22f3b8a99fe
+```
+
+Calling `kpt pkg update` and `kpt pkg update DIR` will perform an update without changing the upstream image name.
+
+At that point, if the `upstream` is an `image:tag` that is to discover the current `image:digest` for tag, otherwise the `upstream` value for `image:digest` is used. In either case, the `upstreamLock` is changed to point at that new `image:digest`. 
+
+The package contents of the old and new `upstreamLock` image digest are fetched to temp folders, and are the basis of the 3-way merge to update the target package.
+
+### `kpt pkg diff`
+
+The `kpt pkg diff` command is identical to `kpt pkg update` in the way that `[PKG_PATH@VERSION]` argument is mapped to OCI concepts.
+
+### Command additions
+
+Although it is possible to create and push an OCI image using a combination of commands like `tar` and `gcrane`, that
+doesn't provide a very complete end to end experience. Because kpt would already be built with the same OCI go module used 
+by `gcrane`, it is not difficult to support additional commands to move pull and push package contents from local folders
+to remote images and back.
+
+### ` kpt pkg push`
+
+```
+Usage: kpt pkg push [DIR] --image={IMAGE[:TAG]}
+  DIR           Folder containing package root Kptfile. Default is current directory.
+  IMAGE[:TAG]   Name of image to push contents onto, with optional TAG to assign to resulting commit. Default TAG is `Latest`
+```
+
+This command will `tar` the contents of the package into a single image layer, and push it into the OCI repository. For
+Google Artifact Registry and Google Container Registry, the current `gcloud auth` SSO credentials are used.
+
+### ` kpt pkg pull`
+
+```
+Usage: kpt pkg pull [DIR] --image={IMAGE[:TAG|@sha256:DIGEST]}
+  DIR                         Destination folder for image contents. Default folder name is the last part of the IMAGE path.
+  IMAGE[:TAG|@sha256:DIGEST]  Name of image to pull contents from, with optional TAG or DIGEST. Default TAG is `Latest`
+```
+
+This command is the reverse of push. An image can be pulled from a repository to a local folder, modified, and pushed
+back to the same location, same location with different TAG, or entirely different location.
+
+### Alternatives to push/pull
+
+There are several ways that pull and push could appear as commands. Those two names are very conventional, but 
+alternatives to consider could be:
+
+### `kpt pkg copy`
```

Comment:
This looks more cleaner to me compared to push and pull commands. However, what does this mean in the context of git ? Do we support this only for oci and not for git ? Should this functionality be included in kpt ? What is the existing alternative to this functionality ?